lib/puppet/network/http/response.rb

```diff
@@ -5,7 +5,18 @@ def initialize(handler, response)
   end
 
   def respond_with(code, type, body)
-    @handler.set_content_type(@response, type)
+    format = Puppet::Network::FormatHandler.format_for(type)
+    mime = format.mime
+
+    if format.supports_charset
+      if body.encoding != Encoding::UTF_8
+        body.encode!(Encoding::UTF_8)
```

Comment:
I'm not sure if you'll ever get into this state (since I don't know what `format.supports_charset` does), but if you have a `BINARY` string here, calling `encode!` will crash.